### PR TITLE
Add explicit outputOrdering for BHJ and SHJ in spark310 shim

### PR DESCRIPTION
**Pull request by andygrove**
Signed-off-by: Andy Grove <andygrove@nvidia.com>

This fixes the regression with running TPC-DS q51 with Spark 3.1.0 and closes https://github.com/NVIDIA/spark-rapids/issues/1203

### CODE REVIEW

**Comment by andygrove** [comment]
build

**Comment by tgravescs** [comment]
title supposed to be 3.1.0?

**Comment by andygrove** [comment]
> title supposed to be 3.1.0?

Yes, thanks.

File: shims/spark310/src/main/scala/com/nvidia/spark/rapids/shims/spark310/GpuBroadcastHashJoinExec.scala [hunk 1]
```diff
@@ -109,6 +109,10 @@ case class GpuBroadcastHashJoinExec(
     "joinTime" -> SQLMetrics.createNanoTimingMetric(sparkContext, "join time"),
     "filterTime" -> SQLMetrics.createNanoTimingMetric(sparkContext, "filter time"))
 
+  // Spark 3.1.0 started introducing ordering guarantees for hash joins but GPU
+  // hash joins do not provide ordering guarantees
+  override def outputOrdering: Seq[SortOrder] = Nil
```
Comment (jlowe) [inline shims/spark310/src/main/scala/com/nvidia/spark/rapids/shims/spark310/GpuBroadcastHashJoinExec.scala:114]: Is it OK for us to replace a node using stricter ordering guarantees with one that is less conformant?  I don't remember whether the ordering constraint pass has occurred before or after the plugin modifies the plan.

Comment (revans2) [reply shims/spark310/src/main/scala/com/nvidia/spark/rapids/shims/spark310/GpuBroadcastHashJoinExec.scala:114]: The ordering fixup happens after we replace the plan.

https://github.com/NVIDIA/spark-rapids/blob/4cb4f5bd01043828953762ebfeec5e6db8c57769/sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOverrides.scala#L2278-L2279

Comment (jlowe) [reply shims/spark310/src/main/scala/com/nvidia/spark/rapids/shims/spark310/GpuBroadcastHashJoinExec.scala:114]: Thanks, @revans2!

**Review by jlowe — APPROVED** [review]

**Review by revans2 — COMMENTED** [review]
This looks OK to me, but I was hoping for perhaps a larger change.  We currently have two copies of `GpuHashJoin` mostly because the parent `HashJoin` we get from Spark changed in 3.1.0.  I would almost rather that we remove inheriting from `HashJoin` and pull in what we need which should hopefully reduce the total code size and insulate us from these types of issues in the future.

**Comment by andygrove** [comment]
OK, I can work on a follow up PR to remove the HashJoin inheritance. I am definitely in favor of this because our hash join operators extend SparkPlan via their base classes as well as via HashJoin and this caused some issues when i was working on this PR.

**Comment by revans2** [comment]
> OK, I can work on a follow up PR to remove the HashJoin inheritance. I am definitely in favor of this because our hash join operators extend SparkPlan via their base classes as well as via HashJoin and this caused some issues when i was working on this PR.

+1

**Pull request merged**